jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java

```diff
@@ -233,9 +183,9 @@ public Content.Chunk read()
             }
             readClosed = current.chunk().isLast();
             if (chunks.isEmpty())
-                l.signal();
+                condition.signal();
         }
-        if (!current.chunk().hasRemaining())
+        if (!current.chunk().canRetain())
```

Comment:
This is the only call to `canRetain` that has any substantive behavior associated with it.  All other calls are in the form:
```java
    if (chunk.canRetain())
        chunk.retain();
````
and there is never an else clause or change to the following logic.   So that raises the question: why don't we just make `retain` a noop for chunks that are not cannot be retained?     The answer is that we need to implement this very specific if statement in this and  only this method!!!!

So at the very least, we should put a great big comment here saying how significant this if statement is.

Also the comment at line 73 above still talks about "terminal".   So it needs to be updated and this mechanism better documented.

I think think we need to take one final step back and decide if `canRetain` really is the right API?  Perhaps we just need an `isConstant` that is used only at this location?

Reply:
More over, we are only using canRetain to pass information between methods of this class, which already has a queue of records.  We can just put that information (to succeed when read) into the object that we queue.

Standby for a PR to this PR....

Reply:
See #9159 that shows t hat `canRetain` is no longer needed

Reply:
I've updated #9159 as per our hangout discussion. I'm OK with it being merged here or this being merged first and mine second.